idx/memory/memory.go

```diff
@@ -639,42 +597,75 @@ func (m *MemoryIdx) Prune(orgId int, oldest time.Time) ([]idx.Archive, error) {
 	return pruned, nil
 }
 
-// filepath.Match doesn't support {} because that's not posix, it's a bashism
-// the easiest way of implementing this extra feature is just expanding single queries
-// that contain these queries into multiple queries, who will be checked separately
-// and whose results will be ORed.
-func expandQueries(query string) []string {
-	queries := []string{query}
+func getMatcher(path string) (func([]string) []string, error) {
+	// Matches everything
+	if path == "*" {
+		return func(children []string) []string {
+			log.Debug("memory-idx: Matching all children")
+			return children
+		}, nil
+	}
+
+	// Convert to regex and match
+	if strings.ContainsAny(path, "*{}[]?") {
+		r, err := regexp.Compile(toRegexp(path))
+		if err != nil {
+			log.Debug("memory-idx: regexp failed to compile. %s - %s", path, err)
+			return nil, err
+		}
+		return func(children []string) []string {
+			matches := make([]string, 0)
+			for _, c := range children {
+				if r.MatchString(c) {
+					log.Debug("memory-idx: %s regex matches %s", c, r.String())
+					matches = append(matches, c)
+				}
+			}
+			return matches
+		}, nil
+	}
+
+	// Match a particular value
+	return func(children []string) []string {
+		for _, c := range children {
+			if c == path {
+				log.Debug("memory-idx: %s matches %s", c, path)
+				return []string{c}
+			}
+		}
+		return []string{}
+	}, nil
+}
 
+func toRegexp(pattern string) string {
 	// as long as we find a { followed by a }, split it up into subqueries, and process
 	// all queries again
 	// we only stop once there are no more queries that still have {..} in them
-	keepLooking := true
-	for keepLooking {
-		expanded := make([]string, 0)
-		keepLooking = false
-		for _, query := range queries {
-			lbrace := strings.Index(query, "{")
-			rbrace := -1
-			if lbrace > -1 {
-				rbrace = strings.Index(query[lbrace:], "}")
-				if rbrace > -1 {
-					rbrace += lbrace
-				}
+	p := pattern
+
+	pos := 0
+	for pos < len(p) {
+
+		lbrace := strings.Index(p[pos:], "{")
+		rbrace := -1
+		if lbrace > -1 {
+			lbrace += pos
+			rbrace = strings.Index(p[lbrace:], "}")
+			if rbrace > -1 {
+				rbrace += lbrace
 			}
+		}
 
-			if lbrace > -1 && rbrace > -1 {
-				keepLooking = true
-				expansion := query[lbrace+1 : rbrace]
-				options := strings.Split(expansion, ",")
-				for _, option := range options {
-					expanded = append(expanded, query[:lbrace]+option+query[rbrace+1:])
-				}
-			} else {
-				expanded = append(expanded, query)
-			}
+		if lbrace == -1 || rbrace == -1 {
+			break
 		}
-		queries = expanded
+		p = p[:lbrace] + "(" + strings.Replace(p[lbrace+1:rbrace], ",", "|", -1) + ")" + p[rbrace+1:]
+		pos = rbrace + 1
+
 	}
-	return queries
+
+	p = strings.Replace(p, "*", ".*", -1)
```

Comment:
one thing i just realized, is that `*` should actually be translated to `[^\.]*` because in graphite world it's only a wildcard within a "node" (where node is dot separated). so we can tell the regex engine to stop looking when it encounters a dot. that will probably give a little performance increase as well.